### PR TITLE
feat: check for cred when updating workflow and remove credential_usage table

### PR DESCRIPTION
**Pull request by krynble**

### CODE REVIEW

File: packages/cli/src/WorkflowHelpers.ts [hunk 1]
```diff
@@ -703,3 +704,56 @@ export function generateFailedExecutionFromError(
 		stoppedAt: new Date(),
 	};
 }
+
+export function validateWorkflowCredentialUsage(
```
Comment (ivov) [inline packages/cli/src/WorkflowHelpers.ts:708]: I rewrote this in order to understand it better. Adding it here just for reference. Untested.

```ts
/**
 * Get all nodes in a workflow where the node credential is not accessible to the user.
 */
function getNodesWithInaccessibleCreds(workflow: WorkflowEntity, userCredIds: string[]) {
	return workflow.nodes.filter((node) => {
		if (!node.credentials) return false;

		return Object.values(node.credentials)
			.map((nodeCred) => nodeCred.id)
			.some((nodeCredId) => nodeCredId !== null && !userCredIds.includes(nodeCredId));
	});
}

const isTamperingAttempt = (prevWorkflowNodeIds: string[], inaccessibleCredNodeId: string) =>
	!prevWorkflowNodeIds.includes(inaccessibleCredNodeId);

/**
 * Validate whether a new version of a workflow has been tampered with.
 */
export function validateNodeCredAccess(
	newWorkflow: WorkflowEntity,
	prevWorkflow: WorkflowEntity,
	userCreds: string[],
) {
	const nodesWithInaccesibleCreds = getNodesWithInaccessibleCreds(newWorkflow, userCreds);

	if (!nodesWithInaccesibleCreds.length) return newWorkflow;

	const prevWorkflowNodeIds = prevWorkflow.nodes.map((node) => node.id);

	nodesWithInaccesibleCreds.forEach((node, index) => {
		if (isTamperingAttempt(prevWorkflowNodeIds, node.id)) {
			throw new Error('Workflow contains new nodes with credentials inaccessible to user');
		}

		// as a precaution, replace node in new worfklow with node in existing workflow
		newWorkflow.nodes[index] = prevWorkflow.nodes.find((prevNode) => prevNode.id === node.id)!;
	});

	return newWorkflow;
}
```

**Review by ivov — COMMENTED** [review]
Tests for all three DBs + PG alt schema passing.

File: packages/cli/src/WorkflowHelpers.ts (hunk 1 above)
Comment (krynble) [reply packages/cli/src/WorkflowHelpers.ts:721]: Thanks, I've improved the code based on your feedback and also fixed an issue I've detected.

**Comment by ivov** [comment]
Re-ran tests for three DBs + PG alt schema and all passing. Pending code review.

File: packages/cli/test/integration/workflows.controller.ee.test.ts [hunk 2]
```diff
@@ -291,7 +284,162 @@ describe('POST /workflows', () => {
 
 		const response = await authAgent(member2).post('/workflows').send(workflow);
 		expect(response.statusCode).toBe(200);
-		const usedCredentials = await testDb.getCredentialUsageInWorkflow(response.body.data.id);
-		expect(usedCredentials).toHaveLength(1);
+	});
+});
+
+describe('PATCH /workflows/:id', () => {
+	it('Should not fail when not touching workflow nodes', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/workflows.controller.ee.test.ts:291]:
```suggestion
	it('Should succeed for unchanged workflow nodes', async () => {
```

File: packages/cli/test/integration/workflows.controller.ee.test.ts [hunk 3]
```diff
@@ -291,7 +284,162 @@ describe('POST /workflows', () => {
...
+		);
+
+		const response = await authAgent(owner).patch(`/workflows/${workflow.id}`).send({
+			name: 'new name',
+		});
+
+		expect(response.statusCode).toBe(200);
+	});
+
+	it('Should allow owner to add node with credential without access', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/workflows.controller.ee.test.ts:324]:
```suggestion
	it('Should allow owner to add node containing credential not accessible to owner', async () => {
```

File: packages/cli/test/integration/workflows.controller.ee.test.ts [hunk 4]
```diff
@@ -291,7 +284,162 @@ describe('POST /workflows', () => {
...
+							},
+						},
+					},
+				],
+			});
+
+		expect(response.statusCode).toBe(200);
+	});
+
+	it('Should not allow member to add node containing credential without access', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/workflows.controller.ee.test.ts:355]:
```suggestion
	it('Should prevent member from adding node containing credential inaccessible to member', async () => {
```

File: packages/cli/test/integration/workflows.controller.ee.test.ts [hunk 5]
```diff
@@ -291,7 +284,162 @@ describe('POST /workflows', () => {
...
+							},
+						},
+					},
+				],
+			});
+
+		expect(response.statusCode).toBe(400);
+	});
+
+	it('Should override existing node with read only version while allowing to save', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/workflows.controller.ee.test.ts:395]:
```suggestion
	it('Should succeed but prevent modifying nodes that are read-only for the requester', async () => {
```

File: packages/cli/src/WorkflowHelpers.ts [hunk 6]
```diff
@@ -703,3 +704,68 @@ export function generateFailedExecutionFromError(
 		stoppedAt: new Date(),
 	};
 }
+
+/** * Get all nodes in a workflow where the node credential is not accessible to the user. */
```
Comment (ivov) [inline packages/cli/src/WorkflowHelpers.ts:708]:
```suggestion
/** Get all nodes in a workflow where the node credential is not accessible to the user. */
```
Else it's parsed as a bullet point:
image

File: packages/cli/test/unit/WorkflowHelpers.test.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	});
+});
+
+function generateCredentialEntity(credentialId: string) {
+	const credentialEntity = new CredentialsEntity();
+	credentialEntity.id = parseInt(credentialId, 10);
+	return credentialEntity;
+}
+
+function getValidWorkflow(options?: {
```
Comment (ivov) [inline packages/cli/test/unit/WorkflowHelpers.test.ts:153]: If there is no invalid variant, maybe just `getWorkflow`?

File: packages/cli/test/unit/WorkflowHelpers.test.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+		test('Should return an element for a node with two credentials and mixed access', () => {
+			const workflow = getValidWorkflow({ addNodeWithTwoCreds: true });
+			const nodesWithInaccessibleCreds = getNodesWithInaccessibleCreds(workflow, [
+				SECOND_CREDENTIAL_ID,
+			]);
+			expect(nodesWithInaccessibleCreds).toHaveLength(1);
+		});
+
+		test('Should return one element for a workflows with two nodes and two credentials', () => {
```
Comment (ivov) [inline packages/cli/test/unit/WorkflowHelpers.test.ts:53]:
```suggestion
		test('Should return one node for a workflow with two nodes and two credentials', () => {
```
Also elsewhere.

File: packages/cli/test/unit/WorkflowHelpers.test.ts [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+		});
+	});
+
+	describe('validateWorkflowCredentialUsage', () => {
+		it('Should throw error saving a workflow using credential without access', () => {
+			const newWorkflowVersion = getValidWorkflow({ addNodeWithOneCred: true });
+			const previousWorkflowVersion = getValidWorkflow();
+			expect(() => {
+				validateWorkflowCredentialUsage(newWorkflowVersion, previousWorkflowVersion, []);
+			}).toThrowError();
```
Comment (ivov) [inline packages/cli/test/unit/WorkflowHelpers.test.ts:104]:
```suggestion
			}).toThrow();
```
According to the docs `toThrowError()` is intended for checking a specific error. Odd that the arg is optional though.

Also elsewhere.

File: packages/cli/src/workflows/workflows.controller.ee.ts [hunk 10]
```diff
@@ -11,7 +18,7 @@ import { SharedWorkflow } from '../databases/entities/SharedWorkflow';
 import { LoggerProxy } from 'n8n-workflow';
 import * as TagHelpers from '../TagHelpers';
 import { EECredentialsService as EECredentials } from '../credentials/credentials.service.ee';
-import { CredentialUsage } from '../databases/entities/CredentialUsage';
+import { FindManyOptions } from 'typeorm';
```
Comment (ivov) [inline packages/cli/src/workflows/workflows.controller.ee.ts:21]:
```suggestion
import type { FindManyOptions } from 'typeorm';
```

File: packages/cli/src/workflows/workflows.controller.ee.ts [hunk 11]
```diff
@@ -202,3 +186,163 @@ EEWorkflowController.post(
 		};
 	}),
 );
+
+EEWorkflowController.patch(
+	'/:id(\\d+)',
```
Comment (ivov) [inline packages/cli/src/workflows/workflows.controller.ee.ts:191]: `(\\d+)` is absent in the free version. Should it be in both?

Comment (ivov) [reply packages/cli/src/workflows/workflows.controller.ee.ts:191]: 90% of this endpoint is identical to the free version.
- Should we abstract away the duplication and import it from both versions if the license allows?
- Should we mark the unique part of the paid version with `// enterprise-edition-start` and `// enterprise-edition-end` to be able to see this section more clearly if the license allows? I had to go through line by line side by side to find the differences.

If the license disallows this, we have a tension between maintainability and the license.

**Review by ivov — COMMENTED** [review]
For testing we are mostly relying on status codes. When doing tests for Ben, as an added precaution he often asked to make assertions about the resulting DB state to ensure that the entity was not mutated. Up to you if you think this should be added.

File: packages/cli/test/unit/WorkflowHelpers.test.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+	});
+});
+
+function generateCredentialEntity(credentialId: string) {
+	const credentialEntity = new CredentialsEntity();
+	credentialEntity.id = parseInt(credentialId, 10);
+	return credentialEntity;
+}
+
+function getWorkflow(options?: {
```
Comment (ivov) [inline packages/cli/test/unit/WorkflowHelpers.test.ts:158]: We already have a `makeWorkflow()` at `/packages/cli/test/integration/shared/utils.ts` Maybe we can consolidate? Only if quick and easy, not a blocker.

**Review by ivov — DISMISSED** [review]
LGTM

File: packages/cli/test/unit/WorkflowHelpers.test.ts (hunk 12 above)
Comment (krynble) [reply packages/cli/test/unit/WorkflowHelpers.test.ts:158]: This one was actually inspired by that one but they are slightly different.

Here we're focusing more on credentials so I've left it separate. We can merge later if needed.

**Review by ivov — DISMISSED** [review]

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with `n8n@0.200.0`

**Pull request merged**